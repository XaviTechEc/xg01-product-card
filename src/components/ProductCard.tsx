import React, { createContext } from 'react';

import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import {
	InitialValues,
	IOnChangeArgs,
	IProduct,
	IProductContextProps,
	ProductCardHandlers
} from '../interfaces/Product';

// import { ProductTitle, ProductButtons, ProductImage } from './';

export const ProductContext = createContext({} as IProductContextProps);
export const { Provider } = ProductContext;

export type ProductCardProps = {
	product: IProduct;
	// children?: React.ReactElement | React.ReactElement[];
	children: (args: ProductCardHandlers) => JSX.Element;
	className?: string;
	style?: React.CSSProperties;
	onChange?: (args: IOnChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
};

export const ProductCard = ({
	children,
	product,
	className,
	style,
	onChange,
	value,
	initialValues
}: ProductCardProps) => {
	const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues
		});

	//* Values available in the provider
	const values = { counter, increaseBy, product, maxCount };

	return (
		<Provider value={values}>
			<div
				className={`${styles.productCard} ${className}`}
				style={style}
			>
				{
					//* === FUNTION CHILD ===
					children({
						count: counter,
						isMaxCountReached,
						maxCount,
						product,
						increaseBy,
						reset
					})
				}
			</div>
		</Provider>
	);
};

// ProductCard.Title = ProductTitle;
// ProductCard.Image = ProductImage;
// ProductCard.Buttons = ProductButtons;
