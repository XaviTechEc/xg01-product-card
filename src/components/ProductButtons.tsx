import React from 'react';
import { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from '../styles/styles.module.css';

export type ProductButtonsProps = {
  className?: string;
};

export const ProductButtons = ({ className }: ProductButtonsProps) => {
  // TODO: Extraer maxCount
  const { counter, increaseBy, maxCount } = useContext(ProductContext);

  // TODO: isMaxReached = useCallback, [counter, maxCount]
  // TRUE si counter === maxCount ELSE FALSE

  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  const handleAdd = () => {
    if (isMaxReached()) return;
    increaseBy(1);
  };

  return (
    <div className={`${styles.buttonsContainer} ${className}`}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        disabled={isMaxReached()}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
        onClick={handleAdd}
      >
        +
      </button>
    </div>
  );
};
