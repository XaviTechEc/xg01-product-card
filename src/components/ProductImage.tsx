import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';

import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export type ProductImageProps = {
  img?: string;
  className?: string;
};

export const ProductImage = ({ img = '', className }: ProductImageProps) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;
  imgToShow = img ? img : product.img ? product.img : noImage;

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Product Img"
    />
  );
};
