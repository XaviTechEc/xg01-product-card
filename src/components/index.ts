import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductCardHOCProps } from '../interfaces/Product';

//* === IMPORTS ===
import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

//* === EXPORTS ===
export { ProductTitle } from './ProductTitle';
export { ProductImage } from './ProductImage';
export { ProductButtons } from './ProductButtons';
// export * from './ProductCard';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
