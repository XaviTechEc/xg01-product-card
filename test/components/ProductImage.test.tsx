import React from 'react';
import renderer from 'react-test-renderer';

import { product2 } from '../data/products';
import { ProductImage, ProductCard } from '../../src/components';

describe('ProductImage', () => {
  test('Must show external img src', () => {
    const wrapper = renderer.create(
      <ProductImage img="https://source.unsplash.com/random" />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Must show img with the img of a product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {() => {
          return <ProductImage />;
        }}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
