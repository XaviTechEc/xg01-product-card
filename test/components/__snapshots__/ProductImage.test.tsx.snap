// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Must show external img src 1`] = `
<img
  alt="Product Img"
  className="productImg undefined"
  src="https://source.unsplash.com/random"
/>
`;

exports[`ProductImage Must show img with the img of a product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product Img"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
